winlocalprocessspawner/winlocalprocessspawner.py

```diff
@@ -44,14 +86,8 @@ def get_env(self):
         token = None
 
         cmd.append(sys.executable)
-        #in virtualenv, exe is in Scripts folder already vs main environment where exe is a python root
-        if hasattr(sys, 'real_prefix'):
-            #in virtual env
-            py_scripts_dir = os.path.dirname(sys.executable)
-        else:
-            #in main env
-            py_scripts_dir = os.path.join(os.path.dirname(sys.executable), 'Scripts')
-        cmd.append(os.path.join(py_scripts_dir, "jupyterhub-singleuser"))
+        
+        cmd.append(self.get_spawner_cmd())

```

Comment:
Sorry for not being clearer on my last post, but my thinking is that the Spawner shouldn't have to do any path manipulation at all. We can just use the configured command:

cmd.extend(self.cmd)

The python executable and python Scripts directory should already be on the PATH. Adding them to the PATH is the responsibility of the calling process. If you want to start the single-user server in a different python environment, then you can use the cmd configuration token to specify the full path to the single-user server.

Reply:
While testing this, I found that jupyter-singleuser is not an exe, so it can't be found by the spawner, even if its on the PATH. I created a PR to create exe's, instead of scripts on jupyterhub during installation https://github.com/jupyterhub/jupyterhub/pull/2313. 

Once that's in, then the only code needded to build the cmd will be:

cmd.extend(self.cmd)
cmd.extend(self.get_args())

As long as the python Scripts is on the PATH, then things should work.

Reply:
So that was part of it.

Another part is that spawner.cmd is allowed to be either a string, or a list which represents a string, so that error has to be corrected (or at least thats how I interpret "Provide either a string or a list containing the path to the startup script command."). So a user could enter "path\\to\\my\\script" or ["path","to","my","script"]. If you do cmd.extend on the latter, I would assume you'd get "python.exe path to my script" rather than the desired "python.exe path\to\my\script".

The other part is that it wasn't able to find the file. In my log I see:
```
Spawning 'C:\tools\jh\vjh\Scripts\python.exe' jupyterhub-singleuser --port=3015
C:\tools\jh\vjh\Scripts\python.exe: can't open file 'jupyterhub-singleuser': [Errno 2] No such file or directory
```
When I try to run that directly from cmd.exe, python can't locate the script. It appears as though python isn't trying to do a path extension, even though the file is in the path ie: echo %PATH% ... "C:\tools\jh\vjh\Scripts;..."

The same is true even if I use the relative path to the working directory:
```
Spawning 'C:\tools\jh\vjh\Scripts\python.exe' 'vjh\Scripts\jupyterhub-singleuser'
C:\tools\jh\vjh\Scripts\python.exe: can't open file 'vjh\Scripts\jupyterhub-singleuser': [Errno 2] No such file or directory
```
I'm assuming this is due to some of the stuff I saw manipulating the working directory. The above command works just fine from cmd.exe, but not from jupyterhub.

So, in order to get it to run, I have to convert to an absolute path. But that isn't the default -- jupyterlab's default is the relative path, so I added the code which converts over to absolute automatically in an attempt to make it more friendly.

Reply:
I believe the documentation is wrong...if you see the base Spwaner at https://github.com/jupyterhub/jupyterhub/blob/master/jupyterhub/spawner.py it does cmd.extend(self.cmd), so yeah, I don't think that's going to work with a string. This is a bug to report/fix on Jupyterhub...since this is a me too Spawner, I think its fine to follow suit, even if we only work with lists (my guess is that on core Jupyterhub the fix is going to be to just adjust the documentation).

Regarding the "[Errno 2] No such file or directory" error, that is the issue that I fixed on the commit I referenced on my previous comment (https://github.com/jupyterhub/jupyterhub/pull/2313). To get a working system, you need to install jupyterhub from a fixed branch. So, from your system you need to run:

python -m pip install --upgrade git+https://github.com/adelcast/jupyterhub@dev/adelcast/entry_points

The setup script will create executables, instead of just scripts, on the Scripts folder. The jupyterhub-singleuser exe will now be locatable at run time.

Reply:
Ah, I thought building the exe just made it so we didn't have to put python.exe in the front, not that it would support the path fix. Makes sense.

Reply:
I'll look at it some more later, but when I imported your version using the pip command specified, it breaks the instance. About 5 months ago spawner.py in the core repo was changed to import pwd which is a unix-only library
`  File "c:\tools\jh\vjh\lib\site-packages\jupyterhub\spawner.py", line 17, in <module>
    import pwd
ModuleNotFoundError: No module named 'pwd'`

Reply:
ah yes, that's another upstream error (I was going to submit a patch to fix it). That import should be in a function, instead of as a global import (Windows doesn't have pwd support). You can comment that import line for now.

Reply:
Fixed on https://github.com/jupyterhub/jupyterhub/pull/2315

Reply:
all righty, I've retested with that pwd line commented out and can confirm it loads correctly (a5a7276). I guess the next step is waiting for those pulls to be merged into main jupyter.